src/box/box.cc

```diff
@@ -5381,10 +5421,20 @@ bootstrap(bool *is_bootstrap_leader)
 
 	while (true) {
 		box_restart_replication();
-		master = replicaset_find_join_master();
-		bootstrap_leader_uuid = master == NULL ? INSTANCE_UUID :
-							 master->uuid;
-		box_broadcast_ballot();
+		/*
+		 * If the bootstrap strategy is 'supervised' and
+		 * it is known that the given instance is the
+		 * bootstrap leader, just bootstrap the database.
+		 *
+		 * Otherwise, find who is the bootstrap leader.
+		 */
+		if (bootstrap_strategy != BOOTSTRAP_STRATEGY_SUPERVISED ||
+		    !am_i_bootstrap_leader) {
+			master = replicaset_find_join_master();
+			bootstrap_leader_uuid = master == NULL ? INSTANCE_UUID :
+								 master->uuid;
+			box_broadcast_ballot();
+		}
```

Comment:
I'm not very confident about this part.

First of all, it seems that `replicaset_find_join_master` and underneath `replicaset_find_join_master_*` functions incapsulate all the related logic and I ruin this idea to some extent. However, it is attractive to make `am_i_bootstrap_leader` local for this compilation unit. This way it is easier to spot that it is needed just to pass some information from `box_make_bootstrap_leader` to `bootstrap`.

Next, it seems meaningful to wait till `replication_connect_timeout` or till all the upstreams are connected and verify, whether the remote peers have a database bootstrapped. OTOH, `replicaset_find_join_master` doesn't do it. At least, if there is another instance that advertizes itself as a bootstrap leader, we can choose our own applier object anyway. And we only look for the bootstrap leader flag, not whether the database is bootstraped.

So, it seems that `replicaset_find_join_master` is not about such safety checks. If we decide to add them, they're likely will be separated from the master finding procedure.

Taking all that, I've implemented in the given way. However, I'm not very in context about this code and if I missed something, please, let me know.

Reply:
I'm not fully in context, but you removed those lines:
```c
		master = replicaset_find_join_master();
		bootstrap_leader_uuid = master == NULL ? INSTANCE_UUID :
							 master->uuid;
		box_broadcast_ballot();
```
why didn't you move them under an `else` statement?

Reply:
It seems, you were confused by the identation change. The `-b` diff looks so:

```diff
@@ -5381,10 +5421,20 @@ bootstrap(bool *is_bootstrap_leader)

        while (true) {
                box_restart_replication();
+               /*
+                * If the bootstrap strategy is 'supervised' and
+                * it is known that the given instance is the
+                * bootstrap leader, just bootstrap the database.
+                *
+                * Otherwise, find who is the bootstrap leader.
+                */
+               if (bootstrap_strategy != BOOTSTRAP_STRATEGY_SUPERVISED ||
+                   !am_i_bootstrap_leader) {
                        master = replicaset_find_join_master();
                        bootstrap_leader_uuid = master == NULL ? INSTANCE_UUID :
                                                                 master->uuid;
                        box_broadcast_ballot();
+               }

                if (master == NULL ||
                    tt_uuid_is_equal(&master->uuid, &INSTANCE_UUID)) {
```

Reply:
Yeap, you're correct)

Reply:
> Next, it seems meaningful to wait till replication_connect_timeout or till all the upstreams are connected

But we already do that in `box_restart_replication` right before searching for `bootstrap_leader`.

> verify, whether the remote peers have a database bootstrapped

Why do we want to do that? If replica advertises itself as bootstrap leader in ballot, we'll bootstrap from it.

Reply:
> > Next, it seems meaningful to wait till replication_connect_timeout or till all the upstreams are connected
>
> But we already do that in `box_restart_replication` right before searching for `bootstrap_leader`.

It seems, we don't (in this bootstrap strategy).

https://github.com/tarantool/tarantool/blob/89cd7f43fd37e7afa459bf1a8f0e5159085c73ae/src/box/replication.cc#L1010-L1015

> > verify, whether the remote peers have a database bootstrapped
>
> Why do we want to do that? If replica advertises itself as bootstrap leader in ballot, we'll bootstrap from it.

As a safety measure. For example, look on this logic:

https://github.com/tarantool/tarantool/blob/89cd7f43fd37e7afa459bf1a8f0e5159085c73ae/src/box/lua/config/applier/box_cfg.lua#L717-L784

In the `auto` bootstrap strategy we just make the instance RW for a short period of time. In the `supervised` strategy without the safety checks we initialize a new replicaset.

I think that we can pass it over here, but I would return to these checks later.

Reply:
@Totktonada 
It seems you don't need this particular change at all.
I'd rather put the `am_i_bootstrap_leader` logic into `replicaset_find_join_master`. Besides, there's a designated check `replicaset_find_join_master_supervised`. It feels logical to check `am_i_bootstrap_leader` there, rather than in the calling code.

So, I suggest the following incremental diff:
```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index a198d6587f..ec0af2f8e1 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -5421,20 +5421,10 @@ bootstrap(bool *is_bootstrap_leader)
 
        while (true) {
                box_restart_replication();
-               /*
-                * If the bootstrap strategy is 'supervised' and
-                * it is known that the given instance is the
-                * bootstrap leader, just bootstrap the database.
-                *
-                * Otherwise, find who is the bootstrap leader.
-                */
-               if (bootstrap_strategy != BOOTSTRAP_STRATEGY_SUPERVISED ||
-                   !am_i_bootstrap_leader) {
-                       master = replicaset_find_join_master();
-                       bootstrap_leader_uuid = master == NULL ? INSTANCE_UUID :
-                                                                master->uuid;
-                       box_broadcast_ballot();
-               }
+               master = replicaset_find_join_master();
+               bootstrap_leader_uuid = master == NULL ? INSTANCE_UUID :
+                                                        master->uuid;
+               box_broadcast_ballot();
 
                if (master == NULL ||
                    tt_uuid_is_equal(&master->uuid, &INSTANCE_UUID)) {
diff --git a/src/box/replication.cc b/src/box/replication.cc
index ac4d8cbe3e..f4cb7518fa 100644
--- a/src/box/replication.cc
+++ b/src/box/replication.cc
@@ -1509,7 +1509,7 @@ replicaset_find_join_master_supervised(void)
                if (applier_is_bootstrap_leader(applier))
                        leader = replica;
        }
-       if (leader == NULL) {
+       if (leader == NULL && !am_i_bootstrap_leader) {
                tnt_raise(ClientError, ER_CFG, "bootstrap_strategy",
                          "failed to connect to the bootstrap leader");
        }
```

Reply:
@sergepetrenko Are you agree with Nikita? Should I incorporate the logic into `replicaset_find_join_master`?

Reply:
I've pushed an intermediate update that reflects everything we discussed except that I'll add logs a bit later today.

Reply:
> except that I'll add logs a bit later today.

Done.